mods/ts/rules/nod-vehicles.yaml

```diff
@@ -520,6 +522,7 @@ STNK:
 		IsPlayerPalette: true
 		UncloakOn: Attack, Unload, Infiltrate, Demolish, Damage, Heal
 		PauseOnCondition: cloak-force-disabled || empdisable
+		CloakedCondition: cloak-self
```

Comment:
Did you test what happens when you build a Stealth Tank within the radius of a Stealth Generator? I would expect it still plays the cloak sound with this setup. Thinking more about it we might need to restore the old checks, as I'm no longer sure this is doable using conditions.

Reply:
Yes, it is still here. going restore the check but I will change it by getting the same cloak type on the unit.

Reply:
ummm, I meet a issue when get the same cloak type. In fact it is cloak type**s**, like targetables. could we make the "InitialDelay = 0" of TS stealth tank's Cloak to avoid this? 

Still the example: submarine is under the effect of Stealth Generator.

Submerge is a kind of cloak, Stealth is a kind of cloak. In regular logic, being stealth and make the sound of steath field doesn't(or may not) prevent submarine make sound of go under the water.

The use case here is more complicated: only one of the two stealth generators can work. In orignal TS, if what I remeber is correct,  the Stealth Tank doesn't recieve the buff from Stealth Generator: when it get heavily damaged, Stealth Generator cannot cloak it like other heavily damage vehicle. If we want to use the gameplay setting like orinnal TS it is also simple to solve.

In conclusion, we have 3 ideas here:
1. make InitialDelay=0 for stealth tank cloak (I confirm it is working)
2. change back to TS orignal setting: Stealth Generator doesn't affect Stealth Tank
3. add Cloak something to identify the same cloak method, without using CloakTypes (or we use CloakTypes with a not so confusing rule), which may makes the code more confusing.

Reply:
I think the most reasonable and logical way for modders is just not having stealth tank be affected by multiple types of stealth via yaml

Reply:
To me It seems like a bad idea to hardcode the solution

Reply:
If you look at the yaml it makes perfect sense why they cloak twice, so I think it should also be solved via yaml. IMO in this case hardcoding the solution only limits modders

I can see 2 easy ways of fixing this

1. Stealth tank overwrites stealth generator stealth instead of being separate
2. Stealth generator stealth is disabled on stealth tank

Reply:
And one feature creep way. Add a property on stealth called `TriggerCloakOnConditions`. That way you could disable external stealth on stealth tank, but add the external condition to stank cloak. The `TriggerCloakOnConditions` would ignore the internal cloak delay and cloak immediately when any of the set conditions are true `cloakgenerator, crate-cloak`. I'm not sure if we can set conditions this way though

Reply:
What is the problem with using CloakTypes?

Reply:
> What is the problem with using CloakTypes?
CloakTypes cannot be used as a standard that tell OpenRA some Cloak traits belong to one kind of cloak, which have the same cloak&uncloak sound/visual effect and have some special rule when playing the sound/visual effect. For example, Stealth Tank cloak and Cloak Generator cloak.

if we want to solve this problem perfectly, we need another property in CloakInfo like CloakClass/CloakKind or something, restore the check but this check is only effective on the same CloakClass/CloakKind. 

To make it more readable I want to rename CloakTypes into DetectableTypes, and make another CloakType to differ the same cloak.

Reply:
> To make it more readable I want to rename CloakTypes into DetectableTypes, and make another CloakType to differ the same cloak.

That sounds like a good idea. I suggest `DetectionTypes` (because they're used for detection, and the types themselves are not detectable).

Reply:
And it is done

Reply:
> Untested but looks good now in theory. Can you add an update rule for the `CloakTypes` -> `DetectionTypes` changes?

I'd love to but I never did the same thing before so could you tell me how to use the rule system?